host.go

```diff
@@ -0,0 +1,205 @@
+		return token.Error()
+	}
+
+	h.mqttClient.Disconnect(uint(h.disconnectTimeout.Milliseconds()))
+	return nil
+}
+
+func (h *HostApplication) initClient() {
+	mqttOpts := mqtt.NewClientOptions()
+	mqttOpts.AddBroker(h.brokerURL)
+	mqttOpts.SetClientID(h.hostID)
+	mqttOpts.SetCleanSession(true)
+	mqttOpts.SetAutoReconnect(true)
+	mqttOpts.SetOrderMatters(true)
+	mqttOpts.SetOnConnectHandler(h.onConnect)
+	mqttOpts.SetReconnectingHandler(h.onReconnect)
+	mqttOpts.SetDefaultPublishHandler(func(_ mqtt.Client, message mqtt.Message) {
+		h.logger.Info("received unexpected message", "topic", message.Topic())
+	})
+
+	h.setLastWill(mqttOpts)
+
+	h.mqttClient = mqtt.NewClient(mqttOpts)
+}
+
+func (h *HostApplication) stateTopic() string {
+	return fmt.Sprintf("%s/%s/%s", sparkplugbNamespace, messageTypeSTATE, h.hostID)
+}
+
+func (h *HostApplication) onConnect(client mqtt.Client) {
+	hostStateTopic := h.stateTopic()
+
+	if token := client.Subscribe(hostStateTopic, byte(1), h.stateHandler); token.Wait() && token.Error() != nil {
+		h.logger.Error(
+			"error subscribing to STATE topic",
+			"topic",
+			hostStateTopic,
+			"error",
+			token.Error().Error(),
+		)
+	}
+
+	// The Sparkplug Host Application MUST publish a
+	// Sparkplug Host Application BIRTH message to the MQTT Server immediately after
+	// successfully subscribing its own spBv1.0/STATE/sparkplug_host_id topic.
+	h.publishOnlineStatus(true, h.lastWillMessageTimestamp)
+
+	client.Subscribe(fmt.Sprintf("%s/+/NBIRTH/+", sparkplugbNamespace), byte(0), h.nodeBirthHandler)
+	client.Subscribe(fmt.Sprintf("%s/+/NDEATH/+", sparkplugbNamespace), byte(0), h.nodeDeathHandler)
+}
+
+func (h *HostApplication) onReconnect(_ mqtt.Client, options *mqtt.ClientOptions) {
+	h.setLastWill(options)
+}
+
+func (h *HostApplication) stateHandler(client mqtt.Client, message mqtt.Message) {
+	// the STATE topic uses QOS 1, so we need to make sure we ACK the message
+	defer message.Ack()
+
+	var status statusPayload
+	if err := json.Unmarshal(message.Payload(), &status); err != nil {
+		h.logger.Error("failed to unmarshal STATE payload", "error", err.Error())
+		return
+	}
+
+	// if at any point the Host Application is delivered a STATE message on
+	// its own Host Application ID with an online value of false, it MUST immediately republish its STATE
+	// message to the same MQTT Server with an online value of true and the timestamp set to the same
+	// value that was used for the timestamp in its own prior MQTT CONNECT packet Will Message
+	// payload
+	if !status.Online {
+		// this runs on a separate goroutine due to paho's mqtt recommendation
+		// that "callback functions must not block or call functions within this
+		// package that may block (e.g. Publish)"
+		go h.publishOnlineStatus(true, h.lastWillMessageTimestamp)
+	}
+}
+
+func (h *HostApplication) setLastWill(mqttOpts *mqtt.ClientOptions) {
+	h.lastWillMessageTimestamp = time.Now().UTC()
+
+	payload, _ := json.Marshal(statusPayload{
```

Comment:
Is it ok to hardcode this to Json?
When do we use proto encoding instead?

Reply:
The Death and Birth Certificate Payload of the host APP MUST be JSON UTF-8 data according to the spec: https://www.eclipse.org/tahu/spec/sparkplug_spec.pdf#payloads_b_state

That's the only message that is not proto as far as I've seen